src/libraries/Common/tests/System/Net/Http/Http2LoopbackConnection.cs

```diff
@@ -558,7 +558,9 @@ public async Task<byte[]> ReadBodyAsync(bool expectEndOfStream = false)
             do
             {
                 frame = await ReadFrameAsync(_timeout).ConfigureAwait(false);
-                if (frame == null && expectEndOfStream)
+
+                // Check if it was a zero-byte read or we got a RST_STREAM with the CANCEL code.
+                if ((frame == null || frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8) && expectEndOfStream)
```

Comment:
Just a suggestion. I had to lookup operator precedence of `&&` and `||` to make sure I understand the condition correctly.
```suggestion
                if ((frame == null || (frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8) && expectEndOfStream))
```

Reply:
I see the confusion, but then we need to add even more parenthesis because there are more `||` and `&&` operations. Don't you think it could make the code less readable?

```suggestion
                if ((frame == null || ((frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8)) && expectEndOfStream))
```

Reply:
Now I'm fully confused what the right order of evaluation is...

Reply:
`&&` takes [precedence](https://docs.microsoft.com/en-us/cpp/c-language/precedence-and-order-of-evaluation?view=msvc-160) over `||` so the order is
1. frame == null
OR
2. frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8 && expectEndOfStream

Reply:
What about this?
```suggestion
                if ((frame == null 
                     || (frame.Type == FrameType.RstStream
                         && ((RstStreamFrame)frame).ErrorCode == 0x8))
                    && expectEndOfStream)
```
Or for simplicity, move `expectedEndOfStream` to the beginning:
```suggestion
                if (expectEndOfStream
                    && (frame == null 
                            || (frame.Type == FrameType.RstStream
                                && ((RstStreamFrame)frame).ErrorCode == 0x8)))
```

Reply:
I can write it as follows if you wish
```suggestion
                if ((frame == null || (frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8) && expectEndOfStream))
```

Reply:
But originally it was:
```frame == null && expectEndOfStream```

What you're posting is will be true if `frame is null` regardless of `expectEndOfStream`, is that what we want?

Reply:
@karelz `expectEndOfStream` cannot be moved forward because the right logic is "frame is null OR (frame type is RstStream AND ErrorCode equals 0x8 AND expectedEndOfStream is true)

Reply:
Yes, you are right I looked at a wrong place. The essence of the change is to handle `frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8` the same way as `frame == null`.

Reply:
Will it be better this way?
```suggestion
                if (expectEndOfStream
                && (frame == null || frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8))
```

Reply:
I'm taking my comment back, please leave it as-is. We're not making the code more readable, it's not worth the hassle.

Reply:
I think we need to change the code to make order of || and && explicit. It is not something anyone wants to keep looking up and I bet that 99% of developers don't remember it. I will rather survive more parenthesis than mind-blowing thinking about || and && precedence.

Reply:
This way then?
```suggestion
                if (expectEndOfStream
                && (frame == null || (frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8)))
```

Reply:
IMO at minimum this:
```
```suggestion
if (expectEndOfStream
    && (frame == null || (frame.Type == FrameType.RstStream && ((RstStreamFrame)frame).ErrorCode == 0x8)))
```

Reply:
Done

Reply:
Haha, that's what I suggested initially (with the exception of moving `expectEndOfStream` at the beginning) :smile: